File: src/eye_tracking/static/measurements.mjs
// Schema v1, matching measurements.py. Input is always the original camera image.
export const EYES = {
  left: { corners: [362, 263], upper: 386, lower: 374, iris: 473 },
  right: { corners: [33, 133], upper: 159, lower: 145, iris: 468 },
};
export const clamp = (x, lo = 0, hi = 1) => Math.max(lo, Math.min(hi, x));
export function hasEyeDetail(points, width, height) {
  if (!points || points.length < 478) return false;
  const spans = Object.values(EYES).map(({ corners: [a, b] }) =>
    Math.hypot(
      (points[b].x - points[a].x) * width,
      (points[b].y - points[a].y) * height,
    ),
  );
  return (
    Math.min(...spans) >= 22 && Math.min(...spans) / Math.max(...spans) > 0.65
  );
}
export function measureEye(points, indices, width, height, score = null) {
  const pixel = (i) => [points[i].x * width, points[i].y * height];
  const [a, b] = indices.corners.map(pixel);
  const dx = b[0] - a[0],
    dy = b[1] - a[1],
    span = Math.hypot(dx, dy);
  if (span < 1e-6) return null;
  const ux = dx / span,
    uy = dy / span;
  const local = (i) => {
    const p = pixel(i),
      x = p[0] - (a[0] + b[0]) / 2,
      y = p[1] - (a[1] + b[1]) / 2;
    return [(x * ux + y * uy) / span, (-x * uy + y * ux) / span];
  };
  const upper = local(indices.upper)[1],
    lower = local(indices.lower)[1];
  const aperture = Math.max(0, lower - upper),
    blink = score === null ? null : clamp(score);
  return {
    iris_image: [points[indices.iris].x, points[indices.iris].y],
    iris_local: local(indices.iris),
    upper_lid: upper,
    lower_lid: lower,
    aperture,
    blink_score: blink,
    openness: blink === null ? null : 1 - blink,
    gaze_valid: aperture > 0.03 && (blink === null || blink < 0.5),
  };
}
export function makePacket(result, timestamp, width, height) {
  const packet = {
    schema_version: 1,
    timestamp_ms: Math.floor(timestamp),
    image_size: [width, height],
    face_present: false,
    face_center: null,
    face_transform: null,
    eyes: null,
  };
  const points = result.faceLandmarks?.[0];
  if (!points || points.length < 478) return packet;
  const scores = Object.fromEntries(
    (result.faceBlendshapes?.[0]?.categories ?? []).map((x) => [
      x.categoryName,
      x.score,
    ]),
  );
  packet.face_present = true;
  packet.eyes = Object.fromEntries(
    Object.entries(EYES).map(([side, indices]) => [
      side,
      measureEye(
        points,
        indices,
        width,
        height,
        scores[`eyeBlink${side[0].toUpperCase()}${side.slice(1)}`] ?? null,
      ),
    ]),
  );
  packet.face_center = ["x", "y"].map((axis) => {
    const values = points.map((p) => p[axis]);
    return (Math.min(...values) + Math.max(...values)) / 2;
  });
  const matrix = result.facialTransformationMatrixes?.[0];
  if (matrix) {
    // MediaPipe JS exposes packed column-major data; Python exposes nested rows.
    packet.face_transform = Array.from({ length: matrix.rows }, (_, row) =>
      Array.from(
        { length: matrix.columns },
        (_, column) => matrix.data[column * matrix.rows + row],
      ),
    );
  }
  return packet;
}

File: src/eye_tracking/measurements.py
"""Pure geometry in a roll-aligned eye frame; no camera or MediaPipe dependency."""

from dataclasses import asdict, dataclass
from math import hypot

# Anatomical sides of the person, using unmirrored MediaPipe input.
# corners are ordered toward increasing image x on a frontal face.
EYES = {
    "left": {"corners": (362, 263), "upper": 386, "lower": 374, "iris": 473},
    "right": {"corners": (33, 133), "upper": 159, "lower": 145, "iris": 468},
}


@dataclass(frozen=True)
class EyeState:
    iris_image: tuple[float, float]
    iris_local: tuple[float, float]
    upper_lid: float
    lower_lid: float
    aperture: float
    blink_score: float | None
    openness: float | None
    gaze_valid: bool


def measure_eye(landmarks, indices, width, height, blink_score=None):
    """Local x/y and lid offsets are in eye-width units from the corner midpoint."""

    def pixel(index):
        point = landmarks[index]
        return point.x * width, point.y * height

    a, b = (pixel(i) for i in indices["corners"])
    dx, dy = b[0] - a[0], b[1] - a[1]
    span = hypot(dx, dy)
    if span < 1e-6:
        return None
    ux, uy = dx / span, dy / span
    midpoint = ((a[0] + b[0]) / 2, (a[1] + b[1]) / 2)

    def local(index):
        px, py = pixel(index)
        x, y = px - midpoint[0], py - midpoint[1]
        return ((x * ux + y * uy) / span, (-x * uy + y * ux) / span)

    upper = local(indices["upper"])[1]
    lower = local(indices["lower"])[1]
    aperture = max(0.0, lower - upper)
    iris = landmarks[indices["iris"]]
    blink = None if blink_score is None else max(0.0, min(1.0, blink_score))
    return EyeState(
        iris_image=(iris.x, iris.y),
        iris_local=local(indices["iris"]),
        upper_lid=upper,
        lower_lid=lower,
        aperture=aperture,
        blink_score=blink,
        openness=None if blink is None else 1.0 - blink,
        gaze_valid=aperture > 0.03 and (blink is None or blink < 0.5),
    )


def make_packet(result, timestamp_ms, width, height):
    packet = {
        "schema_version": 1,
        "timestamp_ms": timestamp_ms,
        "image_size": [width, height],
        "face_present": False,
        "face_center": None,
        "face_transform": None,
        "eyes": None,
    }
    if not result.face_landmarks or len(result.face_landmarks[0]) < 478:
        return packet
    landmarks = result.face_landmarks[0]
    scores = {
        item.category_name: item.score
        for item in (result.face_blendshapes[0] if result.face_blendshapes else [])
    }
    eyes = {}
    for side, indices in EYES.items():
        state = measure_eye(
            landmarks, indices, width, height, scores.get(f"eyeBlink{side.title()}")
        )
        eyes[side] = asdict(state) if state else None
    packet.update(
        face_present=True,
        face_center=[
            (min(p.x for p in landmarks) + max(p.x for p in landmarks)) / 2,
            (min(p.y for p in landmarks) + max(p.y for p in landmarks)) / 2,
        ],
        eyes=eyes,
    )
    if len(result.facial_transformation_matrixes):
        packet["face_transform"] = result.facial_transformation_matrixes[0].tolist()
    return packet
